File: packages/mathlib/src/mathlib.cpp
#include "mathlib.hpp"
#include <cmath>
#include <stdexcept>

namespace mathlib {

double add(double a, double b) {
    return a + b;
}

double sub(double a, double b) {
    return a - b;
}

double mul(double a, double b) {
    return a * b;
}

double div(double a, double b) {
    if (std::fpclassify(b) == FP_ZERO) {
        throw std::invalid_argument("mathlib::div: division by zero");
    }
    return a / b;
}

double abs(double n) {
    return n < 0.0 ? -n : n;
}

} // namespace mathlib

File: packages/mathlib/src/mathlib.lua
-- mathlib: Simple math utilities for REX projects

local mathlib = {}

-- Returns the sum of two numbers
function mathlib.add(a, b)
  return a + b
end

-- Returns the difference of two numbers
function mathlib.sub(a, b)
  return a - b
end

-- Returns the product of two numbers
function mathlib.mul(a, b)
  return a * b
end

-- Returns the quotient of two numbers; errors on division by zero
function mathlib.div(a, b)
  if b == 0 then
    error("mathlib.div: division by zero")
  end
  return a / b
end

-- Returns the absolute value of a number
function mathlib.abs(n)
  if n < 0 then return -n end
  return n
end

return mathlib
